Azure Data Factory CI starter workflow

fedeoliv opened the pull request:
This PR introduces the Azure Data Factory CI starter workflow.

NinadKavimandan commented on ci/properties/azure-data-factory.properties.json:
```diff
@@ -0,0 +1,7 @@
+    "creator": "Microsoft Azure",
+    "description": "Build and validate Azure Data Factory resources",
+    "iconName": "azure-data-factory",
+    "categories": ["Continuous integration", "Azure", "Data Factory"]
```
Are there any criteria to detect if a repository has Azure Data Factory resources? `categories` field is used to examine how relevant a template is to any concerned repository.

fedeoliv replied:
@NinadKavimandan one criteria could be detecting if a repository has a `package.json` file that represents the Data Factory utilities package:

```json
{
    "scripts":{
        "build":"node node_modules/@microsoft/azure-data-factory-utilities/lib/index"
    },
    "dependencies":{
        "@microsoft/azure-data-factory-utilities":"^0.1.5"
    }
} 
```

fedeoliv replied:
Is there a built-in way to check if certain files exists in the repository? Any suggestion?

fedeoliv replied:
One potential approach I've seen is contributing to [languages.yml](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml) on linguist to look for a `package.json` file that contains the `@microsoft/azure-data-factory-utilities` dependency. Not sure if there's a way to make this validation on linguist.

fedeoliv replied:
I've updated the `categories` section to look for `.json` files.

fedeoliv replied:
One potential improvement in the future is contributing to [languages.yml](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml) on linguist or using another way to look for a `.json` file that contains the `@microsoft/azure-data-factory-utilities` keyword.

NinadKavimandan commented:
@fedeoliv LGTM. On the bubbling up front, we use repo analysis in the backend of GitHub now, and templates are recommended based on these `tech_stacks`.

These are the `tech_stacks` we currently detect - https://github.com/github-starter-workflows/repo-analysis-partner/blob/main/tech_stacks.yml

The detection engine is based on `Linguist` detection, so the criteria of detection are inline with `languages.yml`.
You can raise an issue to request detection of Azure Data Factory as a `tech_stack`.

NinadKavimandan commented:
> 
```
{
    "scripts":{
        "build":"node node_modules/@microsoft/azure-data-factory-utilities/lib/index"
    },
    "dependencies":{
        "@microsoft/azure-data-factory-utilities":"^0.1.5"
    }
} ```

@fedeoliv This can be presented as a regex and that can be the detection criteria for `Azure Data Factory`.

fedeoliv commented:
@NinadKavimandan  thanks for the suggestion. I opened an issue to support ADF: https://github.com/github-starter-workflows/repo-analysis-partner/issues/1

NinadKavimandan commented:
@fedeoliv is there any working example of this workflow?

fedeoliv commented:
@NinadKavimandan I have one example here: https://github.com/fedeoliv/data-factory-workflows-demo.
I've been working on a step-by-step CI/CD tutorial for Azure docs (hopefully adding this starter template on the doc soon 😉)

fedeoliv commented:
This is the Azure doc I'm working on: https://github.com/fedeoliv/azure-docs-pr/blob/data-factory-ci-cd-github/articles/data-factory/continuous-integration-delivery-github-actions.md

NinadKavimandan commented:
@fedeoliv Thanks for pointing to the example!
Is having `package.json` with the aforementioned script not a requirement for the workflow to run successfully? I didn't see a `package.json` in your example repo either.

fedeoliv commented:
@NinadKavimandan I made some refactoring on new versions of [data-factory-validate](https://github.com/Azure/data-factory-validate-action) and [data-factory-export](https://github.com/Azure/data-factory-export-action) to simplify even more the GH actions usage. The `package.json` is not needed anymore as it's now embedded in a Dockerfile.

NinadKavimandan commented:
@fedeoliv in that case, will `package.json` play any part?

fedeoliv commented:
@NinadKavimandan not anymore. I was thinking if it's possible to do the following: If the repo has a [publish_config.json](https://docs.microsoft.com/en-us/azure/data-factory/source-control#configure-publishing-settings) file, use the `tech_stack` to detect and suggest ADF workflow templates. Otherwise, let the developer select manually on the list of CI/CD starters.

What do you think? Is it viable?

NinadKavimandan commented:
@fedeoliv detection criteria cannot be based on the directory name. Is the presence of `publish_config.json` good enough criteria?

fedeoliv commented:
@NinadKavimandan that's a good one for sure :)

NinadKavimandan commented:
@fedeoliv cool then! We can take either of these two routes now -

- Merge this PR now, I will add support for `Azure Data Factory` tech_stack and update categories for your template props

or

- You can update the categories to include `Azure Data Factory` right now and merge, whenever `Azure Data Factory` tech_stack is added to repo analysis, this template will be bubbled up.

What would you prefer? Either seems fine to me.

NinadKavimandan commented:
Also
If there's any repo with `publish_config.json` handy, please share.

fedeoliv commented:
@NinadKavimandan I updated the `.properties` to include the `Azure Data Factory` tech_stack. Also, I added a `publish_config.json` sample on my [data-factory-workflows-demo](https://github.com/fedeoliv/data-factory-workflows-demo) repo. Thank you! 😉

NinadKavimandan commented:
@fedeoliv thanks for pointing to the example. This is really helpful!
Will keep you posted about the developments on the repo analysis side.

Going ahead with the merge.

NinadKavimandan commented:
Looks like I don't have the option to merge 😢 
@ashwinsangem can you take a look and merge this PR?
